recipe/patch_yaml_utils.py

```diff
@@ -0,0 +1,310 @@
+            subk = k[:-3]
+            keep = keep and _fnmatch_str_or_list(record[subk], v)
+            if not keep:
+                break
+        elif k == "has_depends":
+            if not isinstance(v, list):
+                v = [v]
+            keep = keep and all(
+                any(fnmatch.fnmatch(dep, _v) for dep in record.get("depends", []))
+                for _v in v
+            )
+            if not keep:
+                break
+        elif k in record:
+            if k == "version":
+                keep = keep and parse_version(record[k]) == parse_version(v)
+            else:
+                keep = keep and fnmatch.fnmatch(str(record[k]), str(v))
+            if not keep:
+                break
+        else:
+            raise KeyError("Unrecognized 'where' key '%s'!" % k)
+
+    return keep
+
+
+def _extract_track_feature(record, feature_name):
+    features = record.get("track_features", "").split()
+    features.remove(feature_name)
+    return " ".join(features) or None
+
+
+def _replace_pin(old_pin, new_pin, deps, record, target="depends"):
+    """Replace an exact pin with a new one. deps and target must match."""
+    if target not in ("depends", "constrains"):
+        raise ValueError
+    if old_pin in deps:
+        i = record[target].index(old_pin)
+        record[target][i] = new_pin
+
+
+def _rename_dependency(fn, record, old_name, new_name):
+    depends = record["depends"]
+    dep_idx = next(
+        (q for q, dep in enumerate(depends) if dep.split(" ")[0] == old_name), None
+    )
+    if dep_idx is not None:
+        parts = depends[dep_idx].split(" ")
+        remainder = (" " + " ".join(parts[1:])) if len(parts) > 1 else ""
+        depends[dep_idx] = new_name + remainder
+        record["depends"] = depends
+
+
+def pad_list(lst, num):
+    if len(lst) >= num:
+        return lst
+    return lst + ["0"] * (num - len(lst))
+
+
+def get_upper_bound(version, max_pin):
+    num_x = max_pin.count("x")
+    ver = pad_list(version.split("."), num_x)
+    ver[num_x:] = ["0"] * (len(ver) - num_x)
+    ver[num_x - 1] = str(int(ver[num_x - 1]) + 1)
+    return ".".join(ver)
+
+
+def _relax_exact(fn, record, fix_dep, max_pin=None):
+    depends = record.get("depends", ())
+    dep_idx = next(
+        (q for q, dep in enumerate(depends) if dep.split(" ")[0] == fix_dep), None
+    )
+    if dep_idx is not None:
+        dep_parts = depends[dep_idx].split(" ")
+        if len(dep_parts) == 3 and not any(
+            dep_parts[1].startswith(op) for op in OPERATORS
+        ):
+            if max_pin is not None:
+                upper_bound = get_upper_bound(dep_parts[1], max_pin) + "a0"
+                depends[dep_idx] = "{} >={},<{}".format(*dep_parts[:2], upper_bound)
+            else:
+                depends[dep_idx] = "{} >={}".format(*dep_parts[:2])
+            record["depends"] = depends
+
+
+CB_PIN_REGEX = re.compile(r"^>=(?P<lower>\d(\.\d+)*a?),<(?P<upper>\d(\.\d+)*)a0$")
```

Comment:
I know this is existing code that moved from the other module, but I wonder how tricky it'd be to use the `MatchSpec` here directly to avoid this kind of regex.

Reply:
I am definitely going to punt on this one, but feel free to try!

Reply:
Given how minimal our testing is here and how essential this data is, I am already nervous about big changes.